src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1290,4 +1308,109 @@ protected void operationProcessed() {
                 threadPool, indexingService, indexSettingsService, warmer, store, deletionPolicy, translog, mergePolicyProvider, mergeScheduler,
                 mapperAnalyzer, similarityService.similarity(), codecService, failedEngineListener, translogRecoveryPerformer);
     }
+
+    private class WriteOperationsCounter extends AbstractRefCounted {
+
+        // for waiting on the counter to get to 0
+        CountDownLatch closingLatch = new CountDownLatch(1);
+
+        WriteOperationsCounter(String name) {
+            super(name);
+        }
+
+        @Override
+        protected void closeInternal() {
+            assert getNumInFlightOperations() == -1;
+            closingLatch.countDown();
+        }
+
+        public void reset() {
+            refCount.set(1);
+            closingLatch = new CountDownLatch(1);
+        }
+
+        public boolean tryDecRef() {
+            if (refCount.get() == 0) {
+                return false;
+            } else {
+                decRef();
+                return true;
+            }
+        }
+    }
+
+    /**
+     * @return >= 0 -> number of operations that are currently processed
+     * -1   -> no more write operations accepted until shard unblockWriteOperations() is called
+     */
+    public long getNumInFlightOperations() {
+        return inFlightOperationsCounter.refCount() - 1;
+    }
+
+    /**
+     * Increments ref counter for write operations.
+     *
+     * @return true if the write operations counter could be incremented, false if the counter is at 0
+     */
+    public boolean beforeWriteOperation() {
+        synchronized (inFlightOperationsCounter) {
```

Comment:
this is unclear to me why we need to sync here I also don't really understand the  entire class and why it needs the latch. I think waiting for 0 is a totally different problem an should be solved by a queue or so. There should be a base class counter that is very simple and we shouldn't mix all these usecases here